<center>
    <img src="https://gitlab.com/ibm/skills-network/courses/placeholder101/-/raw/master/labs/module%201/images/IDSNlogo.png" width="300" alt="cognitiveclass.ai logo"  />
</center>


# **Launch Sites Locations Analysis with Folium**


Estimated time needed: **40** minutes


The launch success rate may depend on many factors such as payload mass, orbit type, and so on. It may also depend on the location and proximities of a launch site, i.e., the initial position of rocket trajectories. Finding an optimal location for building a launch site certainly involves many factors and hopefully we could discover some of the factors by analyzing the existing launch site locations.


In the previous exploratory data analysis labs, you have visualized the SpaceX launch dataset using `matplotlib` and `seaborn` and discovered some preliminary correlations between the launch site and success rates. In this lab, you will be performing more interactive visual analytics using `Folium`.


## Objectives


This lab contains the following tasks:

*   **TASK 1:** Mark all launch sites on a map
*   **TASK 2:** Mark the success/failed launches for each site on the map
*   **TASK 3:** Calculate the distances between a launch site to its proximities

After completed the above tasks, you should be able to find some geographical patterns about launch sites.


Let's first import required Python packages for this lab:


In [1]:
!pip3 install folium
!pip3 install wget

In [2]:
import folium
import wget
import pandas as pd

In [3]:
# Import folium MarkerCluster plugin
from folium.plugins import MarkerCluster
# Import folium MousePosition plugin
from folium.plugins import MousePosition
# Import folium DivIcon plugin
from folium.features import DivIcon

If you need to refresh your memory about folium, you may download and refer to this previous folium lab:


[Generating Maps with Python](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module\_3/DV0101EN-3-5-1-Generating-Maps-in-Python-py-v2.0.ipynb)


## Task 1: Mark all launch sites on a map


First, let's try to add each site's location on a map using site's latitude and longitude coordinates


The following dataset with the name `spacex_launch_geo.csv` is an augmented dataset with latitude and longitude added for each site.


In [4]:
# Download and read the `spacex_launch_geo.csv`
spacex_csv_file = wget.download('https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/spacex_launch_geo.csv')
spacex_df=pd.read_csv(spacex_csv_file)
spacex_df.head()

100% [............................................................] 8966 / 8966

Flight Number        Date Time (UTC) Booster Version  Launch Site  \
0              1  2010-06-04   18:45:00  F9 v1.0  B0003  CCAFS LC-40   
1              2  2010-12-08   15:43:00  F9 v1.0  B0004  CCAFS LC-40   
2              3  2012-05-22    7:44:00  F9 v1.0  B0005  CCAFS LC-40   
3              4  2012-10-08    0:35:00  F9 v1.0  B0006  CCAFS LC-40   
4              5  2013-03-01   15:10:00  F9 v1.0  B0007  CCAFS LC-40   

                                             Payload  Payload Mass (kg)  \
0               Dragon Spacecraft Qualification Unit                0.0   
1  Dragon demo flight C1, two CubeSats,  barrel o...                0.0   
2                             Dragon demo flight C2+              525.0   
3                                       SpaceX CRS-1              500.0   
4                                       SpaceX CRS-2              677.0   

       Orbit         Customer Mission Outcome        Landing Outcome  class  \
0        LEO           SpaceX         Success  Failure   (parachute)      0   
1  LEO (ISS)  NASA (COTS) NRO         Success  Failure   (parachute)      0   
2  LEO (ISS)      NASA (COTS)         Success             No attempt      0   
3  LEO (ISS)       NASA (CRS)         Success             No attempt      0   
4  LEO (ISS)       NASA (CRS)         Success             No attempt      0   

         Lat       Long  
0  28.562302 -80.577356  
1  28.562302 -80.577356  
2  28.562302 -80.577356  
3  28.562302 -80.577356  
4  28.562302 -80.577356

Now, you can take a look at what are the coordinates for each site.


In [5]:
# Select relevant sub-columns: `Launch Site`, `Lat(Latitude)`, `Long(Longitude)`, `class`
spacex_df = spacex_df[['Launch Site', 'Lat', 'Long', 'class']]
launch_sites_df = spacex_df.groupby(['Launch Site'], as_index=False).first()
launch_sites_df = launch_sites_df[['Launch Site', 'Lat', 'Long']]
launch_sites_df

Launch Site        Lat        Long
0   CCAFS LC-40  28.562302  -80.577356
1  CCAFS SLC-40  28.563197  -80.576820
2    KSC LC-39A  28.573255  -80.646895
3   VAFB SLC-4E  34.632834 -120.610746

In [6]:
launch_sites_df.columns

Index(['Launch Site', 'Lat', 'Long'], dtype='object')

Above coordinates are just plain numbers that can not give you any intuitive insights about where are those launch sites. If you are very good at geography, you can interpret those numbers directly in your mind. If not, that's fine too. Let's visualize those locations by pinning them on a map.


We first need to create a folium `Map` object, with an initial center location to be NASA Johnson Space Center at Houston, Texas.


In [7]:
# Start location is NASA Johnson Space Center
nasa_coordinate = [29.559684888503615, -95.0830971930759]
site_map = folium.Map(location=nasa_coordinate, zoom_start=10)

We could use `folium.Circle` to add a highlighted circle area with a text label on a specific coordinate. For example,


In [8]:
# Create a blue circle at NASA Johnson Space Center's coordinate with a popup label showing its name
circle = folium.Circle(nasa_coordinate, radius=1000, color='#d35400', fill=True).add_child(folium.Popup('NASA Johnson Space Center'))
# Create a blue circle at NASA Johnson Space Center's coordinate with a icon showing its name
marker = folium.map.Marker(
    nasa_coordinate,
    # Create an icon as a text label
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'NASA JSC',
        )
    )
site_map.add_child(circle)
site_map.add_child(marker)

and you should find a small yellow circle near the city of Houston and you can zoom-in to see a larger circle.


Now, let's add a circle for each launch site in data frame `launch_sites`


*TODO:*  Create and add `folium.Circle` and `folium.Marker` for each launch site on the site map


An example of folium.Circle:


`folium.Circle(coordinate, radius=1000, color='#000000', fill=True).add_child(folium.Popup(...))`


An example of folium.Marker:


`folium.map.Marker(coordinate, icon=DivIcon(icon_size=(20,20),icon_anchor=(0,0), html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'label', ))`


In [9]:
#Just checking what zip() method returns
Latitudes=list(launch_sites_df.Lat)
Longitudes=list(launch_sites_df.Long)
Launch_site=list(launch_sites_df['Launch Site'])
for lat, lang, launch_site in zip(Latitudes, Longitudes,Launch_site):
    print(lat, lang, launch_site)

28.56230196799018 -80.57735647504778 CCAFS LC-40
28.563197177407144 -80.57682003124195 CCAFS SLC-40
28.57325457037815 -80.64689528960382 KSC LC-39A
34.632834161782775 -120.61074553068327 VAFB SLC-4E


In [10]:
# Initialize the map
site_map = folium.Map(location=nasa_coordinate, zoom_start=5)
# For each launch site, add a Circle object based on its coordinate (Lat, Long) values. In addition, add Launch site name as a popup label
Latitudes=list(launch_sites_df.Lat)
Longitudes=list(launch_sites_df.Long)
Launch_site=list(launch_sites_df['Launch Site'])
for lat, lang, launch_site in zip(Latitudes, Longitudes,Launch_site):
    circle=folium.Circle([lat,lang],
                         radius=5,
                         color='#000000', fill=True).add_child(folium.Popup(launch_site)
                         )
    marker=folium.map.Marker([lat, lang],
                             icon=DivIcon(icon_size=(20,20),
                                          icon_anchor=(0,0),
                                          html='<div style="font-size:12;color:#d35400;"><b>%s</b></div>'%launch_site,
                                         )
                            )
    site_map.add_child(circle)
    site_map.add_child(marker)
site_map

The generated map with marked launch sites should look similar to the following:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_markers.png" />
</center>


Now, you can explore the map by zoom-in/out the marked areas
, and try to answer the following questions:

*   Are all launch sites in proximity to the Equator line?
*   Are all launch sites in very close proximity to the coast?

Also please try to explain your findings.


Q1-Are all launch sites in proximity to the Equator line?
Ans- No they are all near coast

Q2- Are all launch sites in very close proximity to the coast?
Ans- Yes

# Task 2: Mark the success/failed launches for each site on the map


Next, let's try to enhance the map by adding the launch outcomes for each site, and see which sites have high success rates.
Recall that data frame spacex_df has detailed launch records, and the `class` column indicates if this launch was successful or not


In [11]:
spacex_df.tail(5)

Launch Site        Lat      Long  class
51  CCAFS SLC-40  28.563197 -80.57682      0
52  CCAFS SLC-40  28.563197 -80.57682      0
53  CCAFS SLC-40  28.563197 -80.57682      0
54  CCAFS SLC-40  28.563197 -80.57682      1
55  CCAFS SLC-40  28.563197 -80.57682      0

Next, let's create markers for all launch records.
If a launch was successful `(class=1)`, then we use a green marker and if a launch was failed, we use a red marker `(class=0)`


Note that a launch only happens in one of the four launch sites, which means many launch records will have the exact same coordinate. Marker clusters can be a good way to simplify a map containing many markers having the same coordinate.


Let's first create a `MarkerCluster` object


In [12]:
marker_cluster = MarkerCluster()


*TODO:* Create a new column in `launch_sites` dataframe called `marker_color` to store the marker colors based on the `class` value


In [13]:

# Apply a function to check the value of `class` column
# If class=1, marker_color value will be green
# If class=0, marker_color value will be red


In [14]:
# Function to assign color to launch outcome
def assign_marker_color(launch_outcome):
    if launch_outcome == 1:
        return 'green'
    else:
        return 'red'
    
spacex_df['marker_color'] = spacex_df['class'].apply(assign_marker_color)
spacex_df.tail(5)

Launch Site        Lat      Long  class marker_color
51  CCAFS SLC-40  28.563197 -80.57682      0          red
52  CCAFS SLC-40  28.563197 -80.57682      0          red
53  CCAFS SLC-40  28.563197 -80.57682      0          red
54  CCAFS SLC-40  28.563197 -80.57682      1        green
55  CCAFS SLC-40  28.563197 -80.57682      0          red

In [15]:
unique_launch_site=spacex_df['Launch Site'].unique()
unique_launch_site

array(['CCAFS LC-40', 'VAFB SLC-4E', 'KSC LC-39A', 'CCAFS SLC-40'],
      dtype=object)

*TODO:* For each launch result in `spacex_df` data frame, add a `folium.Marker` to `marker_cluster`


In [16]:
latitudes = list(spacex_df['Lat'])
longitudes = list(spacex_df['Long'])
labels = list(spacex_df['marker_color'])
launchSites = list(spacex_df['Launch Site'])

from folium import plugins

map_locations = plugins.MarkerCluster().add_to(site_map)

for lat, long, label, launchSite in zip(latitudes , longitudes, labels, launchSites):
        folium.Marker(
                location= [lat, long],
                icon= folium.Icon(color='white', icon_color=label),
                popup= launchSite,
                ).add_to(map_locations)
        
site_map.add_child(map_locations)

# display map
site_map

In [17]:
#another solution for task no.2
latitudes = list(spacex_df['Lat'])
longitudes = list(spacex_df['Long'])
labels = list(spacex_df['marker_color'])
launchSites = list(spacex_df['Launch Site'])

from folium import plugins

map_locations = plugins.MarkerCluster().add_to(site_map)

for lat, long, label, launchSite in zip(latitudes , longitudes, labels, launchSites):
    if label=='red':
        folium.Marker(
            location=[lat, long],
            icon=folium.Icon(color='red'),
            popup=launchSite,
            ).add_to(map_locations)
    else:
         folium.Marker(
            location=[lat, long],
            icon=folium.Icon(color='green'),
            popup=launchSite,
            ).add_to(map_locations)
    
   
    
site_map.add_child(map_locations)

# display map
site_map


Your updated map may look like the following screenshots:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster.png" />
</center>


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_cluster_zoomed.png" />
</center>


From the color-labeled markers in marker clusters, you should be able to easily identify which launch sites have relatively high success rates.


# TASK 3: Calculate the distances between a launch site to its proximities


Next, we need to explore and analyze the proximities of launch sites.


Let's first add a `MousePosition` on the map to get coordinate for a mouse over a point on the map. As such, while you are exploring the map, you can easily find the coordinates of any points of interests (such as railway)


In [18]:
# Add Mouse Position to get the coordinate (Lat, Long) for a mouse over on the map
formatter = "function(num) {return L.Util.formatNum(num, 5);};"
mouse_position = MousePosition(
    position='topright',
    separator=' Long: ',
    empty_string='NaN',
    lng_first=False,
    num_digits=20,
    prefix='Lat:',
    lat_formatter=formatter,
    lng_formatter=formatter,
)

site_map.add_child(mouse_position)
site_map

Now zoom in to a launch site and explore its proximity to see if you can easily find any railway, highway, coastline, etc. Move your mouse to these points and mark down their coordinates (shown on the top-left) in order to the distance to the launch site.


You can calculate the distance between two points on the map based on their `Lat` and `Long` values using the following method:


In [19]:
from math import sin, cos, sqrt, atan2, radians

def calculate_distance(lat1, lon1, lat2, lon2):
    # approximate radius of earth in km
    R = 6373.0

    lat1 = radians(lat1)
    lon1 = radians(lon1)
    lat2 = radians(lat2)
    lon2 = radians(lon2)

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))

    distance = R * c
    return distance

*TODO:* Mark down a point on the closest coastline using MousePosition and calculate the distance between the coastline point and the launch site.


In [20]:
# find coordinate of the closet coastline 
# e.g.,: Lat: 28.56367  Lon: -80.57163
#distance_coastline = calculate_distance(launch_site_lat, launch_site_lon, coastline_lat, coastline_lon)
distance_coastline = calculate_distance(28.56342, -80.57674, 28.56342, -80.56756)
distance_coastline


0.8968114005914352

*TODO:* After obtained its coordinate, create a `folium.Marker` to show the distance


In [21]:
# Create and add a folium.Marker on your selected closest coastline point on the map
# Display the distance between coastline point and launch site using the icon property 
# for example
distance_marker = folium.Marker(
                      [28.56342, -80.56794],
                      icon=DivIcon(
                             icon_size=(20,20),
                             icon_anchor=(0,0),
                             html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance_coastline),
                           )
                     ).add_to(site_map)   
site_map.add_child(distance_marker)

*TODO:* Draw a `PolyLine` between a launch site to the selected coastline point


In [22]:
# Create a `folium.PolyLine` object using the coastline coordinates and launch site coordinate
#lines=folium.PolyLine(locations=coordinates, weight=1)
lines=folium.PolyLine(locations=[ [28.56342, -80.57674],[28.56342, -80.56756]], weight=1)
site_map.add_child(lines)
site_map

Your updated map with distance line should look like the following screenshot:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/launch_site_marker_distance.png" />
</center>


*TODO:* Similarly, you can draw a line betwee a launch site to its closest city, railway, highway, etc. You need to use `MousePosition` to find the their coordinates on the map first


A railway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/railway.png" />
</center>


A highway map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/highway.png" />
</center>


A city map symbol may look like this:


<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/labs/module_3/images/city.png" />
</center>


In [23]:
# find coordinate of the closest coastline
distance_closest_city = calculate_distance(28.56342, -80.57674, 28.10621, -80.63707) 
distance_closest_railway = calculate_distance(28.56342, -80.57674, 28.56356, -80.78876)
distance_closest_highway = calculate_distance(28.56342, -80.57674, 28.56123, -80.57044)
print("distance_closest_city: ", distance_closest_city, "\ndistance_closest_railway: ", distance_closest_railway, "\ndistance_closest_highway: ", distance_closest_highway)


distance_closest_city:  51.197239764404216 
distance_closest_railway:  20.712620443672808 
distance_closest_highway:  0.6619177700851552


In [24]:
city_marker = folium.Marker(
               [28.10621, -80.63707],
               icon=DivIcon(
                       icon_size=(20,20),
                       icon_anchor=(0,0),
                       html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance_closest_city),
               )
            ).add_to(site_map)

railway_marker = folium.Marker(
               [28.56356, -80.78876],
               icon=DivIcon(
                       icon_size=(20,20),
                       icon_anchor=(0,0),
                       html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance_closest_railway),
               )
            ).add_to(site_map)

highway_marker = folium.Marker(
               [28.56123, -80.57044],
               icon=DivIcon(
                       icon_size=(20,20),
                       icon_anchor=(0,0),
                       html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % "{:10.2f} KM".format(distance_closest_highway),
               )
            ).add_to(site_map)

site_map.add_child(city_marker)
site_map.add_child(railway_marker)
site_map.add_child(highway_marker)

site_map




In [25]:
# Create a marker with distance to a closest city, railway, highway, etc.
# Draw a line between the marker to the launch site
line_city= folium.PolyLine(locations= [[28.56342, -80.57674], [28.10621, -80.63707]], weight=1)
line_railway= folium.PolyLine(locations= [[28.56342, -80.57674],[ 28.56356, -80.78876]], weight=1)
line_highway= folium.PolyLine(locations= [[28.56342, -80.57674], [28.56123, -80.57044]], weight=1)

site_map.add_child(line_city)
site_map.add_child(line_railway)
site_map.add_child(line_highway)

site_map


Highway is the closest to the launch site that is 0.66 KM  & Melbourne city is 51.20 KM away for the launch site.
Railway is 20.71 away from the launch site and nearest coastline is 0.90 KM away from the launch site.

After you plot distance lines to the proximities, you can answer the following questions easily:

*   Are launch sites in close proximity to railways?
*   Are launch sites in close proximity to highways?
*   Are launch sites in close proximity to coastline?
*   Do launch sites keep certain distance away from cities?

Also please try to explain your findings.


# Next Steps:

Now you have discovered many interesting insights related to the launch sites' location using folium, in a very interactive way. Next, you will need to build a dashboard using Ploty Dash on detailed launch records.


## Authors


[Yan Luo](https://www.linkedin.com/in/yan-luo-96288783/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2021-01-01)


### Other Contributors


Joseph Santarcangelo


## Change Log


| Date (YYYY-MM-DD) | Version | Changed By | Change Description          |
| ----------------- | ------- | ---------- | --------------------------- |
| 2021-05-26        | 1.0     | Yan        | Created the initial version |


Copyright © 2021 IBM Corporation. All rights reserved.
